src/crimson/os/seastore/cache.cc

```diff
@@ -979,6 +979,15 @@ CachedExtentRef Cache::duplicate_for_write(
   if (i->is_pending())
     return i;
 
+  if (i->is_exist_clean()) {
```

Comment:
Mutating an ObjectDataBlock is not support yet, see the class definition. And the feature is not used by other types of extents, right?

If that's true, what case can invoke this part of logic?

Reply:
other LogicalCachedExtent, like SeastoreNodeExtent or OMapNode.

Reply:
So the split-mutate logic will never be invoked in the current running SeaStore.
I guess there is also no tests to validate this scenario, combined with conflicts and replay.

Do we need worry about the non-existent use case with tests? Things might be more complicated if the extent needs to mutate then split then mutate in the same transaction.

What do you think? @athanatos

Reply:
> I guess there is also no tests to validate this scenario, combined with conflicts and replay.

https://github.com/ceph/ceph/pull/45519/commits/f9f1e60c0929ea2332941a585f8c27aa608ae92a contains replay test.

If there are several transactions try to split the same extent, all these transaction need to retire the target extent first. When one transaction enter prepare phase, it will invalid other transactions.

I can add a new concurrent split test.

> Things might be more complicated if the extent needs to mutate then split then mutate in the same transaction.

for ObjectDataBlock, here is a multiple split test:
https://github.com/ceph/ceph/blob/2ff59e0fa941894bc0baf83296a39888218a610b/src/test/crimson/seastore/test_object_data_handler.cc#L384-L416

split then mutate:
https://github.com/ceph/ceph/pull/45519/commits/f9f1e60c0929ea2332941a585f8c27aa608ae92a has tested.

mutated then split:
I forget to consider this case, current code can not handle it. `map_existing_extent` invoke `ExtentPlacementManager::read` to fill the extent, if the original extent is a mutated extent, it will lose the delta. If we need to support this case, more development is needed.

Reply:
> If we need to support this case, more development is needed.

Seems no need to implement it in this PR right now. But the complete mutate-with-split feature is in the middle state. We might still need asserts as a reminder.

> The use case would be a rados-level transaction that mutates the same logical byte range on an object resulting in a seastore transaction with two writes to the same byte range within an existing allocated extent.

Do you mean that in the future ObjectDataBlock is expected to support mutations and deltas? If so, probably we can fully enable the mutate-with-split feature in that PR.

Reply:
Hmm, ok.  With the ability to split extents, it may never make sense to journal deltas to ObjectDataBlock's, and RBM extents would probably do write-back rather than a split anyway.  I'm ok with deferring this support for later.

Reply:
Also, ensure that the TransactionManager unit tests for writes include a transaction like the one I outline above.

Reply:
I think ObjectDataBlock will probably need to support deltas fairly soon.  For RBM extents, it will be more efficient to journal-ahead a delta and do a write-back.  We probably even want to support deltas for very small writes on ZNS devices.  We could enforce a restriction that a particular extent may not be mapped and mutated in the same transaction, but that strikes me as too subtle.  I'm inclined to go ahead and introduce this support now given that it's got a test.

Reply:
These use cases make sense to me. And the assert looks trivial.

For "very small writes", do you specifically mean the block-unaligned small writes?

Reply:
Yeah